pkg/policy/repository.go

```diff
@@ -0,0 +1,198 @@
+// context and returns the verdict or api.Undecided if no rule matches. The
+// policy repository mutex must be held.
+func (p *Repository) CanReachRLocked(ctx *SearchContext) api.Decision {
+	decision := api.Undecided
+
+	for _, r := range p.rules {
+		switch r.canReach(ctx) {
+		// The rule contained a constraint which was not met, this
+		// connection is not allowed
+		case api.Denied:
+			return api.Denied
+
+		// The rule allowed the connection but a later rule may impose
+		// additional constraints, so we store the decision but allow
+		// it to be overwritten by an additional requirement
+		case api.Allowed:
+			decision = api.Allowed
+		}
+	}
+
+	return decision
+}
+
+// AllowsRLocked evaluates the policy repository for the provided search
+// context and return the verdict. If no matching policy allows for the
+// connection, the request will be denied. The policy repository mutex must be
+// held.
+func (p *Repository) AllowsRLocked(ctx *SearchContext) api.Decision {
+	ctx.PolicyTrace("NEW TRACE >> %s\n", ctx.String())
+	decision := api.Denied
+
+	if len(p.rules) == 0 {
+		ctx.PolicyTrace("No policy loaded\n")
+	} else {
+		if p.CanReachRLocked(ctx) == api.Allowed {
+			decision = api.Allowed
+		}
+	}
+
+	ctx.PolicyTrace("END TRACE << verdict: [%s]\n", strings.ToUpper(decision.String()))
+
+	return decision
+}
+
+// ResolveL4Policy resolves the L4 policy for a set of endpoints by searching
+// the policy repository for `PortRule` rules that are attached to a `Rule`
+// where the EndpointSelector matches `ctx.To`. `ctx.From` takes no effect and
+// is ignored in the search.  If multiple `PortRule` rules are found, all rules
+// are merged together. If rules contains overlapping port definitions, the first
+// rule found in the repository takes precedence.
+//
+// TODO: Need better rule merging on conflicting port definitions, concat l7 rules?
+func (p *Repository) ResolveL4Policy(ctx *SearchContext) *L4Policy {
+	result := NewL4Policy()
+
+	ctx.PolicyTrace("Resolving L4 policy for context %+v\n", ctx)
+
+	for _, r := range p.rules {
+		r.resolveL4Policy(ctx, result)
+	}
+
+	ctx.PolicyTrace("Final tree L4 policy: %+v\n", result)
+	return result
+}
+
+// SearchRLocked searches the policy repository for rules which match the
+// specified labels and will return an array of all rules which matched.
+func (p *Repository) SearchRLocked(labels labels.LabelArray) []*api.Rule {
+	result := []*api.Rule{}
+
+	for _, r := range p.rules {
+		if r.Labels.Contains(labels) {
+			result = append(result, &r.Rule)
+		}
+	}
+
+	return result
+}
+
+// Add inserts a rule into the policy repository
+func (p *Repository) Add(r api.Rule) error {
+	p.Mutex.Lock()
+	defer p.Mutex.Unlock()
+
+	realRule := &rule{Rule: r}
+	if err := realRule.validate(); err != nil {
+		return err
+	}
+
+	p.rules = append(p.rules, realRule)
+	return nil
+}
+
+// AddList inserts a rule into the policy repository
+func (p *Repository) AddList(rules []*api.Rule) error {
```

Comment:
I've masked it behind a new `api.Rules` type now so we can easily change this. I don't see a problem with a pointer.